### DISCUSSION TITLE
IndexOutOfRangeException in model.Solve()

### DISCUSSION THREAD

**Discussion by AmitR72** [BriefFiniteElementNet/BriefFiniteElement.Net#147] [Q&A] [answered]
Hi there,
Thanks for this lovely library!

I'm trying to apply the examples in the documentation to a model I draw in Rhino & GrassHopper with a plugin I write in c# using the library.
I get the following error on model.Solve(): IndexOutOfRangeException: 'Index was outside the bounds of the array.'
when I tried Solve_MPC, it didn't show an error, but all forces on the bars (elements) were 0. I tried to use the frame and different loads examples too.
At first, I got an error that the matrix isn't symmetric and not positive definite. And after changing the constraints following this https://github.com/BriefFiniteElementNet/BriefFiniteElement.Net/issues/11, I got over this error, but now getting this one.
![Screenshot 2022-08-31 234926](https://user-images.githubusercontent.com/49387829/187780304-eb89bd0d-57c5-4d8b-b36c-4f2da3f537d1.jpg)

this is my model, where the long vector is ~10 and the shorter is ~5 in length. the other vectors are just marks of the constraints, but as I mentioned, I applied Fixed for all the nodes.

**Comment by epsi1on** [comment]
Hi
Thanks. Note that `Model.Solve()` is deprecated and will be removed. `Model.Solve_MPC()` should be used instead.
It seems you are trying to model a 2d truss, if yes, then maybe all nodes are totally fixed to ground ...

**Comment by epsi1on** [comment, marked as answer]
so it seems you've fixed DoFs more than enough so all loads are directly directed to ground and no element have internal force. Can you please tell me which nodes in your visualized model are connected to the ground?
Is it a 2D truss?
Here is more info on how to track the notposdef exception:
https://github.com/BriefFiniteElementNet/BriefFiniteElement.Net/wiki/How-to-fix-NotPosDef-error
I think it is a little more complicated than is seems.
Thanks

**Reply by AmitR72** [reply]
Hi epsi1on,
Thanks for the help!
It's a 2D truss. it actually lies in the XY plain. After I got error, I tries to make all the nodes as Released and that resulted with 0. But then, with the help of your first answer, I tried to put all nodes but the 2 on top (left and middle) as fixed. and these 2 as Released and it worked. I got some results. still got a couple of bars with an internal force of zero, not sure if it makes sense.
I actually trying to build kind of a generic plugin that will give a result for every model. So it's ok if in some cases if the input is not stable, the result will be kind of an error. This model is just an example I'm trying. For sure it looks like that for every model, I will need to get the proper constraints.

**Reply by epsi1on** [reply]
- if it is a truss (only axial strength, not bending on any element), then `Dz`, `Rx`, `Ry`, `Rz` on ALL NODES need to be fixed (partial fix)
- at least some nodes need to be totally fixed to ground to make model stable. (try to google `Truss Stability`)

If these things are applied, probably the `NotPositiveDefinite` exception will be disappeared.

**Reply by AmitR72** [reply]
Thank you! It worked!
now, another question: if I use GetInternalForceAt(0) on each bar after Solve_MPC(), I get:
curve at (0, 5, 0) --- (5, 5, 0) with internal force: {F: -12.5, 0, 0, M: 0, 0, 0}
curve at (0, 5, 0) --- (10, 5, 0) with internal force: {F: 17.5, 0, 0, M: 0, 0, 0}
curve at (5, 5, 0) --- (7.5, 0, 0) with internal force: {F: -16.7705098312485, 0, 0, M: 0, 0, 0}
curve at (0, 5, 0) --- (2.5, 0, 0) with internal force: {F: -11.180339887499, 0, 0, M: 0, 0, 0} 
curve at (5, 5, 0) --- (2.5, 0, 0) with internal force: {F: 11.180339887499, 0, 0, M: 0, 0, 0} 
curve at (10, 5, 0) --- (7.5, 0, 0) with internal force: {F: -39.1311896062463, 0, 0, M: 0, 0, 0} 
curve at (2.5, 0, 0) --- (7.5, 0, 0) with internal force: {F: -10, 0, 0, M: 0, 0, 0} 

Why do I have only forces at x? Not what I would expcet. And, doest negative force means the bar is in strech, while positive force means it's in stress? I want to get this info for each bar, stress/strech.
Thanks again!

**Reply by epsi1on** [reply]
`BarElement.GetInternalForceAt(0)` wil give you the internal force at the middle of element.
For truss elements we have only axial force (in local coordination system, it is along  element's `X` axis)
About the sign, as long as i remember, positive should meamn stretch and negative mean compress.

**Reply by epsi1on** [reply]
Also if the comment is solution to your problem, please use the button (`mark as answer`) to mark it as answer.
Thanks

**Reply by AmitR72** [reply]
Sorry for the many questions...
If I want to apply kind of a G force on a structure.. Should I just use the Dead loadCase with CoordinationSystem.Global for each bar? including the vertical bars? I want to make sure that the self-weight of the vertical bars are also being taking in account in the forces caculations, and I'm not sure that the uniformLoad suits this. (pretty similar to the 4th example here https://www.codeproject.com/Articles/794983/Finite-Element-Method-Programming-in-Csharp only without xy forces and z forces with self-mass of the verticals bars)
Thanks!

**Reply by epsi1on** [reply]
I think it would work, UniformLoad, with `Vector.K` direction and `DeadLoad` load case and `CoordinationSystem.Global` direction on every bar (even verticals). Anyways need to be tested.

There are also 7 other well documented examples on documentation. They could help too. [link](https://bfenet.readthedocs.io/en/latest/example/index.html)

Example list is in the left panel:
![image](https://user-images.githubusercontent.com/7839141/188794533-43d1bf04-3bea-47ce-a419-11366462dc35.png)

**Comment by epsi1on** [comment]
inactivity
